htlcswitch/link.go

```diff
@@ -2419,6 +2564,71 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 
 }
 
+// drivePendingStfuSends checks to see if we owe any Stfu messages to our peer.
+// If we do then we will send them and record that send. We can owe our peer an
+// Stfu either if we have a pending request to quiesce the channel or our peer
+// has sent us an Stfu to which we haven't yet responded.
+func (l *channelLink) drivePendingStfuSends() {
+	if !l.quiescer.oweStfu() {
+		return
+	}
+
+	l.log.Debug("stfu owed")
+
+	handleErr := func(err error) {
+		l.log.Error(err.Error())
+		l.stfuFailf(err.Error())
+	}
+
+	localOnRemote := l.channel.NumLocalUpdatesPendingOnRemote()
+	localOnLocal := l.channel.NumLocalUpdatesPendingOnLocal()
+	if localOnRemote == 0 && localOnLocal == 0 {
+		l.log.Debug("all local updates synced, sending stfu")
+		oStfu, err := l.quiescer.sendStfu()
+		if err != nil {
+			handleErr(err)
+			return
+		}
+		oStfu.WhenSome(func(stfu lnwire.Stfu) {
+			err := l.cfg.Peer.SendMessage(false, &stfu)
+			if err != nil {
+				handleErr(err)
+				return
+			}
+			l.tryResolveQuiescenceRequests()
+		})
+	}
+}
+
+// tryResolveQuiescenceRequests checks whether the quiescence protocol has
+// successfully completed. If it has, it completes the outstanding asynchronous
+// request, if there is one.
+func (l *channelLink) tryResolveQuiescenceRequests() {
```

Comment:
Some of the logic here and in `drivePendingStfuSends` seems like it could be better homed in the `quiescer` than the link because that's where all of this state information lives. 

If we flip things around, I think we can present the link with a cleaner API - something  like: 

- `init(resp <-chan)`: registers the intent to quiesce and provides the notification channel
- `receiveStfu`: reports that a stfu message has been received
- `sendStfu(sendMsg func()) bool, error`: sends a stfu if owed, returning a bool if the message was sent

Reply:
We can't because stfu sends have to be synchronized with normal link operations. At the end of the day the quiescer is just a bucket for state and state transition rules. It should not be performing IO actions on its own. I'd rather use the quiescer as a means of making decisions, not as a means of performing actions.

There isn't a practical way to separate the quiescer state from the rest of link state given that there are cross-system message sending requirements wherein we cannot send `stfu` in certain link states. I am sympathetic to the idea that there is room for improvement here in terms of code cleanliness, but I think it's important we do not give the quiescer direct access to the message queue.

Reply:
> At the end of the day the quiescer is just a bucket for state and state transition rules.

Is this what we want? I think that a contained `quiescer` that has full understanding of the protocol its driving + a link that just has to state step it forward when link-relevant events happen is more contained than spreading this logic between a state machine and the link. 

Took a shot at this [here](https://github.com/carlaKC/lnd/commit/a4c43cc1d0514416e658a0c56f132e520d2f4440) (while in delayed flight purgatory, so highly imperfect). The API that needs to be exposed to the link is a bit more involved than I was hoping (see your point here), but I do think that it's still an appealing approach.

>  important we do not give the quiescer direct access to the message queue

Why? Quite a common pattern in the codebase to configure standalone subsystems (`quiescer`) with access to drive actions elsewhere (`link`).

Reply:
> Took a shot at this [here](https://github.com/carlaKC/lnd/commit/a4c43cc1d0514416e658a0c56f132e520d2f4440) (while in delayed flight purgatory, so highly imperfect). The API that needs to be exposed to the link is a bit more involved than I was hoping (see your point here), but I do think that it's still an appealing approach.

I took a look and I think this is a perfectly reasonable approach. I definitely see the appeal here too. I also think that @Roasbeef would agree with you here as the approach you took here looks almost *exactly* like the protofsm approach.

> Is this what we want? I think that a contained quiescer that has full understanding of the protocol its driving + a link that just has to state step it forward when link-relevant events happen is more contained than spreading this logic between a state machine and the link.

I agree with the sentiment here. However, I think it is necessary to spread it between the two to some degree. The quiescer, regardless of how we design it, is incapable of directly assessing whether the preconditions for an stfu send are met as this is fundamentally part of normal CSM operations. The API you present makes the link responsible for informing the quiescer when it is allowed to send `stfu`. As such the link code must be aware of some of the protocol details here. The quiescer gets the benefit of reducing this decision to a single continuation that it is given during initialization, but the link had to know what to give it in the first place. The details regarding whether the "TryProgressState" function ought to belong to the link or the quiescer is unclear to me and ultimately hinges on whether or not you want to make the quiescer responsible for sending any sort of messages. 

Ultimately I think your suggestion is to just [CPS encode](https://en.wikipedia.org/wiki/Continuation-passing_style) the logic in this PR, and that seems reasonable enough to me.

> Why? Quite a common pattern in the codebase to configure standalone subsystems (quiescer) with access to drive actions elsewhere (link).

The main reason I said what I did is that I think it's important to separate decisions from actions and that the quiescer was originally designed to just work with the decisions, leaving the actions up to the link where they already are. Your proposed changes make it responsible for both decision and action but you managed to do it in a way that still cleanly separates them and gives the link an easy way to supply its context.

Reply:
> The API you present makes the link responsible for informing the quiescer when it is allowed to send stfu. As such the link code must be aware of some of the protocol details here. 

Indeed, we're never going to get around the link needing to handle initiation because it's the main bus for message. I do however wonder if there's a smarter way to handle `TryProgressState` than the quick-n-dumb direct mapping that I went with. 

eg: Could be nice to have a `quisecer.Update()` method that we just `defer` in `handleUpstreamMessage` (that's a no-op most of the time when we have no state progression/quiescence on the go) to improve separation of concerns between the link and the quiescer.

Reply:
> eg: Could be nice to have a quisecer.Update() method that we just defer in handleUpstreamMessage (that's a no-op most of the time when we have no state progression/quiescence on the go) to improve separation of concerns between the link and the quiescer.

This is evil genius level thinking. I respect the cleverness but I really worry about that making behavior really hard to understand. I think for me to be OK with this I'd have to be able to convince myself that the sensible place to handle this update ticking is always at the end of that function irrespective of which internal branches it takes. It's a good idea and worth considering.

Reply:
> We can't because stfu sends have to be synchronized with normal link operations. At the end of the day the quiescer is just a bucket for state and state transition rules

Can't we use the existing hooks for this? They were added to fix protocol deviation issues in the co-o close flow, and seem well suited for this use case. 

Overall, I think @carlaKC brings up some good points here. At an initial read, the e2e control flow isn't very clear to me. Re `protofsm`, for the new RBF co-op close PR, I was able to utilize the hooks added during the co-op close bug fixes to ensure that my actions were directly inline with state updates in the link (clear to send shutdown, no pending updates, etc).

Reply:
We have to introduce new hooks if we want to do this. The ones we introduced for the shutdown process do not capture the proper nuance for the `stfu` requirements.

Reply:
I've reworked this in full and I think you'll find the new design more pleasing.